scheduler/src/cook/mesos/api.clj

```diff
@@ -58,15 +86,76 @@
    (s/optional-key :ports) [(s/pred zero? 'zero)] ;;TODO add to docs the limited uri/port support
    (s/optional-key :env) {NonEmptyString s/Str}
    (s/optional-key :labels) {NonEmptyString s/Str}
+   (s/optional-key :container) Container
    :cpus PosDouble
    :mem PosDouble
    ;; Make sure the user name is valid. It must begin with a lower case character, end with
    ;; a lower case character or a digit, and has length between 2 to (62 + 2).
    :user (s/both s/Str (s/pred #(re-matches #"\A[a-z][a-z0-9_-]{0,62}[a-z0-9]\z" %) 'lowercase-alphanum?))})
 
+(defn- mk-container-params
+  "Helper for build-container.  Transforms parameters into the datomic schema."
+  [cid params]
+  (when-not (empty? params)
+    {:docker/parameters (mapv (fn [{:keys [key value]}]
+                                {:docker.param/key key
+                                 :docker.param/value value})
+                              params)}))
+
+(defn- mkvolumes
+  "Helper for build-container.  Transforms volumes into the datomic schema."
+  [cid vols]
+  (when-not (empty? vols)
+    (let [vol-maps (mapv (fn [{:keys [container-path host-path mode]}]
+                           (merge (when container-path
+                                    {:container.volume/container-path container-path})
+                                  (when host-path
+                                    {:container.volume/host-path host-path})
+                                  (when mode
+                                    {:container.volume/mode (clojure.string/upper-case mode)})))
+                         vols)]
+      {:container/volumes vol-maps})))
+
+(defn- mk-docker-ports
+  "Helper for build-container.  Transforms port-mappings into the datomic schema."
+  [cid ports]
+  (when-not (empty? ports)
+    (let [port-maps (mapv (fn [{:keys [container-port host-port protocol]}]
+                            (merge (when container-port
+                                     {:docker.portmap/container-port container-port})
+                                   (when host-port
+                                     {:docker.portmap/host-port host-port})
+                                   (when protocol
+                                     {:docker.portmap/protocol (clojure.string/upper-case protocol)})))
+                          ports)]
+      {:docker/port-mapping port-maps})))
+
+(defn- build-container
+  "Helper for submit-jobs, deal with container structure."
+  [id container]
+  (let [container-id (d/tempid :db.part/user)
```

Comment:
(when (and container (= (clojure.string/lower-case (:type container)) "docker))
      (let ...))